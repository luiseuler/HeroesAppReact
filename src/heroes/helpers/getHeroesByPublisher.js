import { heroes } from "../data/heroes"

export const getHeroesByPublisher = (publisher) => {
    const validPublisher = ['DC Comics', 'Marvel Comics'];

    if (!validPublisher.includes(publisher)) {
        throw new Error('El publisher no existe');
    }

    return heroes.filter(hero => hero.publisher === publisher);
}