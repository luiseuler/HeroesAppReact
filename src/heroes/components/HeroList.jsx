import { getHeroesByPublisher } from "../helpers";
import { HeroCard } from "./HeroCard";

export const HeroList = ({ publisher }) => {
    const heroes = getHeroesByPublisher(publisher);

    return (
        <diiv className="row row-cols-1 row-cols-md-3 g-3">
            {
                heroes.map(hero => (<HeroCard key={hero.superhero} {...hero} />))
            }
        </diiv>
    );
}
