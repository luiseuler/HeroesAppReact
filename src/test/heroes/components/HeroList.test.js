import { MemoryRouter } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { HeroList } from "../../../heroes/components/";
import { getHeroesByPublisher } from "../../../heroes/helpers/getHeroesByPublisher";

jest.mock('../../../heroes/helpers/getHeroesByPublisher.js');

describe('Pruebas en <HeroList />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe de hacer match con el snapshot', () => {
        getHeroesByPublisher.mockReturnValue([]);

        const { container } = render(
            <MemoryRouter>
                <HeroList publisher={'DC Comics'} />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    test('debe de llamar getHeroesByPublisher con el parÃ¡metro dado', () => {
        getHeroesByPublisher.mockReturnValue([]);

        render(
            <MemoryRouter>
                <HeroList publisher={'DC Comics'} />
            </MemoryRouter>
        );

        expect(getHeroesByPublisher).toHaveBeenCalledWith('DC Comics');
    });

    test('debe de renderizar los card heroes si hay resultados', () => {
        getHeroesByPublisher.mockReturnValue([
            {
                id: 'dc-batman',
                superhero: 'Batman',
                publisher: 'DC Comics',
                alter_ego: 'Bruce Wayne',
                first_appearance: 'Detective Comics #27',
                characters: 'Bruce Wayne'
            }
        ]);

        const { container } = render(
            <MemoryRouter>
                <HeroList publisher={'DC Comics'} />
            </MemoryRouter>
        );

        const cardsElements = container.querySelectorAll('.card');

        expect(getHeroesByPublisher).toHaveBeenCalledWith('DC Comics');
        expect(cardsElements.length).toBeGreaterThan(0);
    });

    test('debe de NO renderizar los card heroes si NO hay resultados', () => {
        getHeroesByPublisher.mockReturnValue([]);

        const { container } = render(
            <MemoryRouter>
                <HeroList publisher={'DC Comics'} />
            </MemoryRouter>
        );

        const cardsElements = container.querySelectorAll('.card');
        
        expect(getHeroesByPublisher).toHaveBeenCalledWith('DC Comics');
        expect(cardsElements.length).toBe(0);
    });
});