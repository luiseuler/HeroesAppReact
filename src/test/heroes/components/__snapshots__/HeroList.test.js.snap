// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroList /> debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="row row-cols-1 row-cols-md-3 g-3"
  />
</div>
`;
